### ISSUE TITLE
Suggestion: make model path configurable (and DeLFT architecture optional)

### ISSUE THREAD

**Issue by de-code** [kermitt2/grobid#692]
After having merged with upstream, the DeLFT architecture is now added to the model path.

GROBID does seem to handle `null` in some instances, but not a blank string.
Setting `GROBID__DELFT__ARCHITECTURE` to a blank string causes it to try to load models like `header-` (with the added hyphen).

In this case, adding the architecture to the model path causes incompatibility. The main issue I see with it, is that it adds "magic". i.e. looking at the config (`grobid.properties`), I wouldn't guess that it is added to the path or how it is added. `useELMo` is another parameter that is added to the path. Fortunately it can be set to `false`. But if it was set to `true` should the path be `<model>-<architecture>-useELMo` or `<model>-useELMo<architecture>`?. Adding the path to the config, would make that more clear and could add other hyper parameters (or even the overall model type, such as `wapiti`).

Additionally, currently it seems that the `architecture` is assumed to be the same across all models. (I may be wrong).

The downside of per model model path configuration would be additional config parameters, but in this case it seems worth it (and defaults can be provided via `grobid.properties`)

In summary, I would agree that it makes sense to keep each model in a separate folder (I also have them stored in a cloud storage with a quite long name). But building a model path using some of the parameters seems both introducing magic and also limiting.

**Comment by kermitt2** [comment]
Hi @de-code !

I've worked recently on a new single config file for Grobid in yaml, which is in branch `yaml-config`. It introduces full config at model level, so each model can run with different "engine" and "architecture" and there is nothing global anymore (it was confusing yes). When `delft` is set as engine for a model, the `architecture` parameter would be mandatory, so no blank string (it could default to BidLSTM-CRF or just reject the configuration file).

About the path where to store the model, the idea was that it is all managed automatically, so the user would not need to worry about the path (the actual way the path is built does not matter for the user). The model has anyway to be stored somewhere by default, so the motivation is to keep it simple and readable, and I think using model parameters to build the default path is better than a random string and I don't see how to improve that. 

For sure, as improvement we could add a model path parameter in the config file that would be considered with priority to the default build-in path. A URI for a model to be downloaded automatically is of course more work.

**Comment by de-code** [comment]
Thank you for your comment.

I guess there are perhaps three types of model users:

1. users that just want to use the default models provided
(this doesn't really affect them)
2. users that train models via GROBID
Those users may be more concerned with setting the parameters and providing the data.
The model path would only become relevant if one wants to keep a history of models.
3. users that train models outside GROBID (e.g. using DeLFT directly)
In this case all the parameters will generally also be defined outside GROBID and training time.
Now the model path could be different. (I understand you are replicating some of the model path logic in DeLFT)

Option 2. doesn't work so well for me due to the cloud usage. Option 3. works better for me and doesn't require GROBID at training time.

Because I am using Option 3, I generally just have the model path. The architecture would be redundant and could be out of step. i.e. in that case I would prefer not having to specify the architecture (I used `dummy` for now).

--

What I am doing in my Docker entrypoint is downloading models based on environment variables. I mainly do that for evaluation purpose. My "model" under test them mostly consists of the base image and a set of environment variables the specify the model path URLs or change other features. In the [sciencebeam-grobid-biorxiv](https://github.com/elifesciences/sciencebeam-grobid-biorxiv) those models would be copied into the image like the default models.

Automatically loading a model from a remote resource would be great of course. For the way I use things it seems to make things so much easier. I could just give you a link to a model that you update in the config to try a new different version. It also works great for the training data. I could easily just tell the train process to use the biorxiv training data as well as the GROBID default training data for example (by just pointing to the two URLs with the data).

Some things that make it more challenging:
- I am using HTTP (`https://`) and Google Storage (`gs://`) URLs
- Wapiti models may be just a single file but DeLFT models would either point to a cloud directory or a zip

But that is definitely out of scope of this ticket.

**Comment by kermitt2** [comment]
I'll try to add first the "custom" model path in the new grobid config format, and adapt DeLFT accordingly, that does not sound  complicated. 

Remote resources (and independent repo to manage training data and models) are very convincing improvements, but it's much more work and might make things heavier future iteration (more dependencies, more extra mechanisms to maintain).
